enhancements/generic-constraints.md

```diff
@@ -0,0 +1,181 @@
+
+For majority of use cases, the CEL expression should be simple and straight-forward especially with well-constructed examples and guidelines where users can make a few plug-in changes and ready to be used.
+
+The `message` field is introduced to allow specific information from the author who writes the expression to surface as resolution output.
+
+#### CEL language maturity and support
+
+The CEL language is relatively new and not yet at stable release (v1). There are potentially breaking changes that can be introduced.
+
+Migration: The CEL library can be spinned to a specific release to avoid introducing breaking changes if occurs.
+
+The language is being used in another well-known opensource projects including Tekton and Kubernetes ([apiserver/CRD](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2876-crd-validation-expression-language/README.md)). This should improve the confidence on the usage of CEL language and these teams will potentially influence or impose against any major breaking changes that may come to CEL.
+
+If CEL becomes too instable in the future due to breaking changes, it is possible to gradually deprecate the support of the language and introduce another language that is more stable. The constraint syntax allows the support of other languages in the future via `evaluator` field.
+
+## Alternatives
+
+### Other languages choices
+
+While CEL is the selected language to be supoorted in the new OLM constraint type. It is not the intention to limit OLM to only support CEL expression. It possible to support multiple expression languages in the future releases if needed.
+
+Besides CEL, there are other languages that can be used such as:
+
+- [Rego](https://github.com/open-policy-agent/opa/tree/main/rego) (Open Policy Agent)
+- [Expr](https://github.com/antonmedv/expr)
+- [Starlark](https://github.com/bazelbuild/starlark)
+- [Cue](https://github.com/cue-lang/cue)
+
+All of these languages can be supported in constraint type. The `evaluator` field is designed to support multiple evaluators/languages if needed. However, intrducing a new language to the constraint type should be evaluated carefully to ensure there is a real need for it. Providing the support for multiple languages can be overwhelming and potentially creates a fragmented user experiences and unnecessary maintainance effort in a long run.
+
+### Simplified constraint type
+
+The current proposed constraint syntax has most of information nested under `value` field. The complexity of a struct with nested fields can create a bad user experience. It is possible to introduce a constraint type that is specified for CEL
+
+```json=
+{
+  "type": "olm.constraint.cel",
+  "value": {
+    "rule": 'properties.exists(p, p.type == "certified") && properties.exists(p, p.type == "stable")',
+    "message": 'require to have "certified" and "stable" properties',
```

Comment:
This just seems like restating the constraint. Is this actually a field that should be available on in the property metaschema?

```
{
  "type": "proptype",
  "value": <>, 
  "failureMessage": "etc"
}
```

Reply:
I'd buy `"description"` making sense for all properties.

Reply:
Based on the current state of this EP, I think I'd vote to start with `failureMessage` in the root of the `olm.constraint` value struct.

A general `description` and a constraint `failureMessage` seem different enough semantically to me that I don't think we should overload one onto the other.